src/shim.ts

```diff
@@ -6,43 +18,181 @@ import {
  * const df = require("durable-functions");
  *
- * module.exports = df.orchestrator(function*(context) {
- *     // function body
- * });
+ * app.generic('DurableFunctionsOrchestratorJS', {
+ *   trigger: trigger.generic({
+ *       type: 'orchestrationTrigger'
+ *   }),
+ *   handler: df.createOrchestrator(function* (context) {
+ *       // orchestrator body
+ *  })
+ * })
  * ```
  */
-export function orchestrator(
-    fn: (context: IOrchestrationFunctionContext) => Generator<unknown, unknown, any>
-): (
-    context: IOrchestrationFunctionContext,
-    orchestrationTrigger: DurableOrchestrationBindingInfo
-) => Promise<OrchestratorState> {
+export function createOrchestrator(fn: OrchestrationHandler): OrchestrationFunction {
     const listener = new Orchestrator(fn).listen();
-    return async (context, orchestrationTrigger): Promise<OrchestratorState> => {
+
+    return async (
+        context: IOrchestrationFunctionContext,
+        orchestrationTrigger: DurableOrchestrationInput
+    ): Promise<OrchestratorState> => {
         return await listener(context, orchestrationTrigger);
     };
 }
 
-export function entity<T = unknown>(
-    fn: (context: IEntityFunctionContext<T>) => void
-): (
-    context: IEntityFunctionContext<T>,
-    entityTrigger: DurableEntityBindingInfo
-) => Promise<EntityState> {
+/**
+ * Enables an entity handler function to act as a Durable Entity Azure Function.
+ *
+ * @param fn the handler function that should act as a durable entity
+ * @example Register an entity using `app.generic()`
+ * ```javascript
+ * const { app } = require("@azure/functions")
+ * const df = require("durable-functions");
+ *
+ * app.generic('DurableEntity', {
+ *   trigger: df.trigger.entity(),
+ *   handler: df.createEntityFunction(function* (context) {
+ *       // orchestrator body
+ *  })
+ * })
+ * ```
+ */
+export function createEntityFunction<T = unknown>(fn: EntityHandler<T>): EntityFunction<T> {
     const listener = new Entity<T>(fn).listen();
-    return async (context, entityTrigger): Promise<EntityState> => {
+
+    return async (
+        context: IEntityFunctionContext<T>,
+        entityTrigger: DurableEntityBindingInfo
+    ): Promise<EntityState> => {
         return await listener(context, entityTrigger);
     };
 }
+
+/**
+ * Registers a generator function as a Durable Orchestrator for your Function App.
+ *
+ * @param functionName the name of your new durable orchestrator
+ * @param handler the generator function that should act as an orchestrator
+ *
+ * @example Register an orchestrator
+ * ```javascript
+ * const df = require("durable-functions");
+ *
+ * df.orchestration('DurableFunctionsOrchestratorJS', function* (context) {
+ *     // orchestrator body
+ * });
+ * ```
+ */
+export function orchestration(functionName: string, handler: OrchestrationHandler): void {
+    app.generic(functionName, {
+        trigger: trigger.orchestration(),
+        handler: createOrchestrator(handler),
+    });
+}

```

Comment:
Just a thought: After staring at this implementation, I think I finally understand what throws me off about the subtle distinction between `orchestration` and `createOrchestrator`.

The source of my confusion is in that the `handler: ` field in the options to `app.generic` does not accept the `OrchestrationHandler` type. Instead, it accepts an `OrchestratorFunction` type, which is unintuitive to me. I would expect that an `OrchestrationHandler` would be a valid type to pass in to the `handler: ` field here, but it is not. 

Put differently, `createOrchestrator` transforms an `OrchestratorHandler` into an `OrchestratorFunction`, which can then be passed as a `handler: ` to app generic. In my mind, it would be more intuitive if the names were reverse.

Put differently, it would make more sense to me if `createOrchestrator` transformed a type `OrchestratorFunction` into an `OrchestratorHandler` which naturally fits as the type accepted by the `handler: ` field of `app.generic`. Thoughts? @hossam-nasr

Reply:
Discussed offline. `createOrchestrator` and the `OrchestrationFunction` type have been removed from public access so some of this discussion is obsolete now anyway